### PR TITLE
Loading all controllers 

### PR DESCRIPTION
**Pull request by ugocupcic**
Fixes https://github.com/shadow-robot/sr-visualization/issues/35.

Will do a pull request on **sr_config** when this is merge (adding this script to sr_edc_default_controllers.launch).

### CODE REVIEW

File: sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                failed = True
+            if not failed:
+                rospy.loginfo("Loaded "+controller+" sucessfully.")
+
+if __name__ == '__main__':
+    rospy.init_node("load_all_controllers")
+
+    #@TODO replace this sleep with a proper way of identifying when the controllers
+    # can be loaded
+    sleep(6.0)
```
Comment (ugocupcic) [inline sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers]: How can we do that in a clever way?

Comment (markpitchless) [inline sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers]: What is it that is being waited for? I'm guessing a service on the controller manager, just wait for that to be availiable. E.g. list controllers.

Comment (shadowmanos) [inline sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers]: If the firmware in the hand can report when it changed its mode,
we can capture it in the C++ drivers and then expose it in the python lib
too.

Cheers
Manos Nikolaidis

Software Team
Shadow Robot Company. 251 Liverpool Road, London U.K. N1 1LX
Tel: +44 (0) 20 7700 2487 E: manos@shadowrobot.com

On 4 September 2014 15:36, Ugo Cupcic notifications@github.com wrote:

> In
> sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers:
> 
> > +
> > -            try:
> > -                load_controllers = rospy.ServiceProxy('controller_manager/load_controller', LoadController)
> > -                resp1 = load_controllers(controller)
> > -            except rospy.ServiceException:
> > -                rospy.logerr("Failed to load controller "+controller)
> > -                failed = True
> > -            if not failed:
> > -                rospy.loginfo("Loaded "+controller+" sucessfully.")
> >   +
> >   +if **name** == '**main**':
> > -    rospy.init_node("load_all_controllers")
> >   +
> > -    #@TODO replace this sleep with a proper way of identifying when the controllers
> > -    # can be loaded
> > -    sleep(6.0)
> 
> How can we do that in a clever way?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/shadow-robot/sr-ros-interface/pull/154/files#r17116674
> .

## 

---

_Shadow Robot Company Ltd._
251 Liverpool Road, N1 1LX, UK
Registered Number 3308007 (England & Wales)

RoNeX - Building Robots with ROS Made Easy 

## http://www.shadowrobot.com/products/ronex/

**Comment by ugocupcic** [comment]
Found a solution, I need to implement it, wait before merging

File: sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        timeout=args.timeout
+
+        if topic != None:
+            self.wait_for_topic(topic, timeout)
+
+        params = rospy.get_param("")
+
+        unique_keys = []
+        for key in params.keys():
+            if ("sh" in key) and ("controller" in key):
```
Comment (toliver) [inline sr_mechanism_controllers/scripts/sr_mechanism_controllers/load_all_controllers:43]: This is a bold move!! :)

**Comment by ugocupcic** [comment]
Seems like too much of a trouble to go through. Most of this can be done from launch files for easy start up of the hand.

**Comment by markpitchless** [comment]
I was expecting this to just be a change to the launch files!

But then I didn't fully understand what happens with controller config until @toliver explain yesterday. The way we have 2 sets of config for each controller depending on whether we are in PWM or torque mode. Which is seperate issue I think.

**Pull request closed without merging**